ffi/src/transaction/mod.rs

```diff
@@ -0,0 +1,349 @@
+    txn: Handle<ExclusiveTransaction>,
+    commit_info: Handle<ExclusiveEngineData>,
+) -> Handle<ExclusiveTransaction> {
+    let txn = unsafe { txn.into_inner() };
+    let commit_info = unsafe { commit_info.into_inner() };
+    Box::new(txn.with_commit_info(commit_info)).into()
+}
+
+/// Add file metadata to the transaction for files that have been written. The metadata contains
+/// information about files written during the transaction that will be added to the Delta log
+/// during commit.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid handle. Assumes no concurrent txn usage. Consumes
+/// write_metadata.
+#[no_mangle]
+pub unsafe extern "C" fn add_write_metadata(
+    mut txn: Handle<ExclusiveTransaction>,
+    write_metadata: Handle<ExclusiveEngineData>,
+) {
+    let txn = unsafe { txn.as_mut() };
+    let write_metadata = unsafe { write_metadata.into_inner() };
+    txn.add_write_metadata(write_metadata);
+}
+
+/// Attempt to commit a transaction to the table. Returns version number if successful.
+/// Returns error if the commit fails.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid handle. And MUST NOT USE transaction after this
+/// method is called.
+#[no_mangle]
+pub unsafe extern "C" fn commit(
+    txn: Handle<ExclusiveTransaction>,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<u64> {
+    let txn = unsafe { txn.into_inner() };
+    let extern_engine = unsafe { engine.as_ref() };
+    let engine = extern_engine.engine();
+    // FIXME for now just erasing the enum
+    match txn.commit(engine.as_ref()) {
+        Ok(CommitResult::Committed(v)) => Ok(v),
+        Ok(CommitResult::Conflict(_, v)) => Err(delta_kernel::Error::Generic(format!(
+            "commit conflict at version {v}"
+        ))),
+        Err(e) => Err(e),
+    }
+    .into_extern_result(&extern_engine)
+}
+
+#[cfg(test)]
+mod tests {
+    use delta_kernel::schema::{DataType, StructField, StructType};
+
+    use delta_kernel::arrow::array::{Array, ArrayRef, Int32Array, StringArray, StructArray};
+    use delta_kernel::arrow::datatypes::{DataType as ArrowDataType, Field, Schema as ArrowSchema};
+    use delta_kernel::arrow::ffi::to_ffi;
+    use delta_kernel::arrow::record_batch::RecordBatch;
+    use delta_kernel::engine::arrow_data::ArrowEngineData;
+    use delta_kernel::parquet::arrow::arrow_writer::ArrowWriter;
+    use delta_kernel::parquet::file::properties::WriterProperties;
+
+    use delta_kernel_ffi::engine_data::get_engine_data;
+    use delta_kernel_ffi::engine_data::ArrowFFIData;
+
+    use delta_kernel_ffi::tests::{get_default_engine, ok_or_panic};
+
+    use write_context::{free_write_context, get_write_context, get_write_path, get_write_schema};
+    use crate::{free_engine, free_schema, kernel_string_slice};
+
+    use test_utils::{setup_test_tables, test_read};
+
+    use std::sync::Arc;
+
+    use super::*;
+
+    use tempfile::tempdir;
+
+    fn create_arrow_ffi_from_json(
```

Comment:
all these helpers I feel like are pointing to (1) a testing issue - need better facilities for testing all this easily and (2) possibly needing some 'simpler' APIs in general? certainly #997 is going in that direction

Reply:
I think we can take refactoring the tests as a follow-up. That's never going to be a breaking change, so let's prioritize getting this merged.

Reply:
(and sorry should have said in my original comment: that was a bit of an aside, more for us to take on some work in the future not in this PR!)